### PR TITLE
propagate error to message bar for WMS and vector tile

### PR DESCRIPTION
**Pull request by vcloarec**

### CODE REVIEW

File: src/core/network/qgsnetworkaccessmanager.cpp [hunk 1]
```diff
@@ -439,6 +439,9 @@ void QgsNetworkAccessManager::onReplySslErrors( const QList<QSslError> &errors )
   Q_ASSERT( reply );
   Q_ASSERT( reply->manager() == this );
 
+  for ( const QSslError &er : errors )
+    qDebug() << "Error:" << er.error() << "  Error string" << er.errorString();
```
Comment (wonder-sk) [inline src/core/network/qgsnetworkaccessmanager.cpp]: probably can be removed

File: src/core/vectortile/qgsvectortileloader.cpp [hunk 2]
```diff
@@ -130,6 +131,34 @@ void QgsVectorTileLoader::tileReplyFinished()
...
+    QString errorSource = mUri;
+
+    bool notifyError = false;
+
+    switch ( reply->error() )
+    {
+      case QNetworkReply::OperationCanceledError:
+        break;
+      case QNetworkReply::TimeoutError:
+        break;
```
Comment (wonder-sk) [inline src/core/vectortile/qgsvectortileloader.cpp]: Is there a reason why these two error types are explicitly listed, but with no body?

File: src/core/vectortile/qgsvectortileloader.cpp [hunk 3]
```diff
@@ -130,6 +131,34 @@ void QgsVectorTileLoader::tileReplyFinished()
   }
   else
   {
+    QString errorMessage;
+    QString errorSource = mUri;
+
+    bool notifyError = false;
```
Comment (wonder-sk) [inline src/core/vectortile/qgsvectortileloader.cpp]: do we need this variable? maybe we could simply check whether `errorMessage` is empty

File: src/core/vectortile/qgsvectortileloader.cpp [hunk 4]
```diff
@@ -130,6 +131,34 @@ void QgsVectorTileLoader::tileReplyFinished()
...
+      case QNetworkReply::TimeoutError:
+        break;
+      case QNetworkReply::ContentAccessDenied:
+        notifyError = true;
+        errorMessage = tr( "Access denied to \"%1\"" ).arg( mUri );
+        if ( !reply->data().isEmpty() )
+          errorMessage.append( QStringLiteral( "<br>" ) + reply->data() );
+        break;
+      case QNetworkReply::ContentNotFoundError:
+        errorMessage = tr( "Nothing found for %1" ).arg( mUri );
```
Comment (wonder-sk) [inline src/core/vectortile/qgsvectortileloader.cpp]: Hmm I don't think we want to display this as an error to the user... It is common that XYZ tiles in "boring" areas such as oceans respond with http code 404 and it is expected...

File: src/core/vectortile/qgsvectortileloader.cpp [hunk 5]
```diff
@@ -130,6 +131,34 @@ void QgsVectorTileLoader::tileReplyFinished()
...
+        notifyError = true;
+        break;
+      default:
+        break;
+    }
+
+    if ( notifyError )
+      QgsNetworkAccessManager::instance()->notifyError( errorSource, errorMessage );
```
Comment (wonder-sk) [inline src/core/vectortile/qgsvectortileloader.cpp]: Hmm I am not sure if I like the idea that the errors go through NAM... there may be other errors that are not strictly related to network access (e.g. if we fail to decode tiles), so it would be good not to tie error reporting to NAM.

Why not propagate the errors to QgsMapRendererJob::errors() and then have QgisApp show errors in message bar based on those?

**Comment by vcloarec** [comment]
@wonder-sk , PR updated.

I try but I didn't make the aggregation of layers with same message. 
For that we to add a new signal `renderErrorsOccurred( const QString &error, const QList<QgsMapLayer *> &layers )` that the QgisApp receive to make the message for the message bar but ... the only method in QgsMessageBar that accepts `showMore` is `pushMessage()`.

And this method don't handle rich text but only plain text. I try to modify this method to accept also rich text, but that begins to be quite tricky because of the `<br>` or the `\n ` that make the separation between the text and the show more depending of the format of the text.

So I give up for now...

File: src/core/vectortile/qgsvectortileloader.h [hunk 6]
```diff
@@ -91,6 +91,8 @@ class QgsVectorTileLoader : public QObject
     //! Blocks the caller until all asynchronous requests are finished (with a success or a failure)
     void downloadBlocking();
 
+    QString error() const;
```
Comment (wonder-sk) [inline src/core/vectortile/qgsvectortileloader.h:95]: a doxygen comment would be nice

File: src/providers/wms/qgswmsprovider.cpp [hunk 7]
```diff
@@ -4455,10 +4466,14 @@ void QgsWmsTiledImageDownloadHandler::tileReplyFinished()
...
         stat.errors++;
-
         // if we reached timeout, let's try again (e.g. in case of slow connection or slow server)
-        if ( reply->error() == QNetworkReply::TimeoutError )
-          repeatTileRequest( reply->request() );
+        repeatTileRequest( reply->request() );
+
+        if ( reply->error() == QNetworkReply::ContentAccessDenied )
+        {
+          mError = tr( "Access denied for following reason: %2" ).
```
Comment (wonder-sk) [inline src/providers/wms/qgswmsprovider.cpp]: I would shorten this to just "Access denied: %1" (without "for the following reason")

And we should use `%1` instead of `%2` there, right?

File: src/gui/qgsmapcanvas.h [hunk 8]
```diff
@@ -1422,6 +1423,8 @@ class GUI_EXPORT QgsMapCanvas : public QGraphicsView, public QgsExpressionContex
 
     std::unique_ptr< QgsTemporaryCursorOverride > mTemporaryCursorOverride;
 
+    QMap <QString, QDateTime> mRendererErrors;
```
Comment (wonder-sk) [inline src/gui/qgsmapcanvas.h:1433]: It would be nice to document that we are trying not to throw rendering errors on every render, that's why we keep the list of recent errors with time
